WIP: Fix for #243

MarchingCube opened the pull request:
My initial attempt at fixing issues in https://github.com/dadhi/FastExpressionCompiler/issues/243.

While I've managed to fix initial issue when writing tests I figured out one other case that was not tested - using `ref` parameters to then call normal by value methods. This is still causing failures and to be honest - without investing a lot of time I can't really tell what is a correct fix. I've added a comment above a line that might cause it - basically it should emit extra indirect reference load but it doesn't due to a conditional that I don't understand.

MarchingCube commented on src/FastExpressionCompiler/FastExpressionCompiler.cs:
```diff
@@ -2314,7 +2317,7 @@ private static void EmitDereference(ILGenerator il, Type type)
                 if (method != null && method.Name != "op_Implicit" && method.Name != "op_Explicit")
                 {
-                    if (!TryEmit(opExpr, paramExprs, il, ref closure, parent & ~ParentFlags.IgnoreResult | ParentFlags.InstanceCall, 0))
+                    if (!TryEmit(opExpr, paramExprs, il, ref closure, parent & ~ParentFlags.IgnoreResult | ParentFlags.InstanceCall, -1))
```
![image](https://user-images.githubusercontent.com/2588062/75392126-aeb6dd00-58eb-11ea-9d2f-8fd4c3571405.png)

This `0`  caused my other fix to start emitting by reference loads since this tells that first argument should be passed by reference from what I understand. And at least in unit tests this code was calling methods that take arguments by value so tests started to fail.

dadhi replied:
Huh, ok seems a legit fix from the first sight.

MarchingCube commented on src/FastExpressionCompiler/FastExpressionCompiler.cs:
```diff
@@ -2095,14 +2095,16 @@ private static void EmitDefault(Type type, ILGenerator il)
                          (parent & ParentFlags.MemberAccess) != 0);
 
-                    if (closure.LastEmitIsAddress)
+                    if (closure.LastEmitIsAddress || (!paramExpr.IsByRef && isArgumentByRef))
```
This is slightly convoluted way of checking if we are given non-reference parameter but our argument expects a reference. In this case it should load address and not value. Hoping that this is a right place to do this.

dadhi commented:
Thank you, will look Today.

dadhi commented:
Fix seems legit, but it is hard to tell because of AppVeyor failing - likely not relevant to the PR. I will merge it as is and will try to check on my side.
Thank you for your contribution!

MarchingCube commented:
@dadhi AppVeyor might fail since there was one failing test (that's why I've added WIP to the PR) that I mentioned in the description and added `TODO` in code.

dadhi commented:
Ah, will `[Ignore]` it on my side.
